adds tagging support for events (closes #1907)

bryanlalexander opened the pull request:
adds localstack.utils.TaggingService, implements TaggingService for CloudWatch Events

**Please refer to the contribution guidelines in the README when submitting PRs.**

whummer commented on localstack/utils/tagging.py:
```diff
@@ -0,0 +1,24 @@
+
+    def untag_resource(self, arn, tag_names):
+        for name in tag_names:
+            if name in self.tags[arn]:
```
Better:
```
if name in self.tags.get(arn, {}):
    ...
```
(to secure against `KeyError`)

bryanlalexander commented:
@whummer the following changes were made, some of which were not on your list, but we adding integration testing and fixed a couple bugs we found with the downloader.

- moved test_tagging.py to localstack/tests/unit
- fixed inconsistency in test_create_tag (map has no guaranteed key order)
- adds status code and size check to localstack.utils.common.download method
- adds exception logging for localstack.services.install.download_and_extract_with_retry
- fixed deprecated call to assertRegexMatches
- adds integration testing for ListTagsForResource, TagResource, and UntagResource
- adds constants import of APPLICATION_AMZ_JSON_1_1 to events_listener
- moves parsed_data to the foremost if statement in ProxyListenerEvents
- fixed bug in AWSEvents.UntagResource method in ProxyListenerEvents
- updates make_response in events_listener to do the json marshalling
- updates AWSEvents.PutEvents to use make_response in ProxyListenerEvents
- adds updated KeyError-safe check to TaggingService

whummer commented on localstack/utils/common.py:
```diff
@@ -607,8 +611,13 @@ def download(url, path, verify_ssl=True):
             f.flush()
             os.fsync(f)
+        if os.path.getsize(path) == 0:
+            LOG.error('Zero bytes downloaded from %s, retrying' % url)
```
Can we please reduce the log level to `LOG.info` or `LOG.warning`? We're generally only using `error` for fatal errors.

whummer commented on localstack/services/install.py:
```diff
@@ -238,8 +238,9 @@ def download_and_extract():
-    except Exception:
+    except Exception as e:
         # try deleting and re-downloading the zip file
+        LOGGER.error('download_and_extract_with_retry caught exception: %s' % e)
```
Can we combine the two log messages into one (with log level `info` or `warning`)? Something like
```
LOGGER.info('Unable to extract file, re-downloading ZIP archive %s: %s' % (tmp_archive, e))
```

bryanlalexander commented:
@whummer All set, those changes have been made, thanks for the thorough review.

whummer commented:
Awesome, thanks again!
